File: packages/python/langchain_nuggets/middleware/authority_middleware.py
"""NuggetsAuthorityMiddleware — tool-call interception for trust enforcement."""
from __future__ import annotations

import json
import logging
import time
from datetime import datetime, timezone
from typing import Any, Callable, Dict, List, Optional

from langchain_core.messages import ToolMessage

from langchain_nuggets.client.nuggets_api_client import NuggetsApiClient
from langchain_nuggets.middleware.proof import (
    build_proof_artifact,
    hash_parameters,
    hash_result,
)
from langchain_nuggets.middleware.types import (
    ActionContext,
    AuthorityEvaluationRequest,
    AuthorityEvaluationResponse,
    MiddlewareConfig,
    ProofArtifact,
)

logger = logging.getLogger(__name__)


class NuggetsAuthorityMiddleware:
    """Middleware that intercepts LangChain/LangGraph tool calls and enforces
    Nuggets trust primitives: Actor Identity, Authority, Policy, Intent,
    Consent, and Accountability.

    Provides wrap_tool_call and awrap_tool_call functions compatible with
    LangGraph's ToolNode and create_react_agent.

    Usage with ToolNode::

        middleware = NuggetsAuthorityMiddleware(config)
        tool_node = ToolNode(
            tools=tools,
            wrap_tool_call=middleware.wrap_tool_call,
        )
    """

    def __init__(self, config: MiddlewareConfig) -> None:
        self._config = config
        self._client = NuggetsApiClient(
            {
                "api_url": config.api_url,
                "partner_id": config.partner_id,
                "partner_secret": config.partner_secret,
                "ca_cert": config.ca_cert,
                "verify_ssl": config.verify_ssl,
            }
        )
        self._proofs: List[ProofArtifact] = []
        self._on_proof: Optional[Callable[[ProofArtifact], Any]] = config.on_proof

    @property
    def proofs(self) -> List[ProofArtifact]:
        """All proof artifacts emitted during this middleware's lifetime."""
        return list(self._proofs)

    def _build_eval_request(
        self,
        tool_name: str,
        tool_args: Dict[str, Any],
    ) -> AuthorityEvaluationRequest:
        """Construct the authority evaluation request payload."""
        # Resolve intent via callback if configured
        intent = None
        if self._config.intent_resolver is not None:
            intent = self._config.intent_resolver(tool_name, tool_args)

        # Extract target from args if present, otherwise default to tool name
        target = tool_args.get("target", tool_name) if isinstance(tool_args, dict) else tool_name

        return AuthorityEvaluationRequest(
            agent_id=self._config.agent_id,
            controller_id=self._config.controller_id,
            delegation_id=self._config.delegation_id,
            action=ActionContext(
                tool=tool_name,
                target=str(target),
                parameters_hash=hash_parameters(tool_args),
                intent=intent,
                timestamp=datetime.now(timezone.utc).isoformat(),
            ),
        )

    def _make_deny_message(
        self,
        tool_call_id: str,
        tool_name: str,
        response: AuthorityEvaluationResponse,
    ) -> ToolMessage:
        """Create a structured ToolMessage for a DENY decision."""
        content = json.dumps(
            {
                "status": "DENIED",
                "tool": tool_name,
                "reason_code": response.reason_code,
                "proof_id": response.proof_id,
                "message": f"Authority check denied execution of '{tool_name}'"
                + (f": {response.reason_code}" if response.reason_code else ""),
            }
        )
        return ToolMessage(content=content, tool_call_id=tool_call_id)

    def _emit_proof(self, proof: ProofArtifact) -> None:
        """Store proof and invoke callback if configured."""
        self._proofs.append(proof)
        if self._on_proof is not None:
            self._on_proof(proof)

    def wrap_tool_call(self, request: Any, handler: Any) -> Any:
        """Synchronous tool call wrapper compatible with LangGraph ToolNode.

        Evaluates authority before tool execution. On ALLOW, executes the tool
        and emits a proof artifact. On DENY, returns a structured error ToolMessage.
        """
        tool_call = request.tool_call
        tool_name = tool_call["name"]
        tool_args = tool_call["args"]
        tool_call_id = tool_call["id"]

        eval_request = self._build_eval_request(tool_name, tool_args)
        start_time = time.monotonic()

        try:
            raw_response = self._client.post(
                self._config.authority_endpoint,
                eval_request.model_dump(),
            )
            auth_response = AuthorityEvaluationResponse(**raw_response)
        except Exception as exc:
            logger.error("Authority evaluation failed: %s", exc)
            return ToolMessage(
                content=json.dumps(
                    {
                        "status": "ERROR",
                        "tool": tool_name,
                        "message": f"Authority evaluation failed: {exc}",
                    }
                ),
                tool_call_id=tool_call_id,
            )

        if auth_response.decision == "DENY":
            logger.info("DENY: tool=%s reason=%s", tool_name, auth_response.reason_code)
            return self._make_deny_message(tool_call_id, tool_name, auth_response)

        logger.info("ALLOW: tool=%s proof_id=%s", tool_name, auth_response.proof_id)
        result = handler(request)

        result_content = ""
        if isinstance(result, ToolMessage):
            result_content = (
                result.content if isinstance(result.content, str) else str(result.content)
            )

        total_latency = (time.monotonic() - start_time) * 1000
        proof = build_proof_artifact(
            authority_response=auth_response,
            agent_id=self._config.agent_id,
            controller_id=self._config.controller_id,
            delegation_id=self._config.delegation_id,
            tool=tool_name,
            parameters_hash=hash_parameters(tool_args),
            result_hash=hash_result(result_content),
            latency_ms=total_latency,
        )
        self._emit_proof(proof)

        return result

    async def awrap_tool_call(self, request: Any, handler: Any) -> Any:
        """Asynchronous tool call wrapper compatible with LangGraph ToolNode.

        Async variant of wrap_tool_call. Uses the async NuggetsApiClient methods.
        """
        tool_call = request.tool_call
        tool_name = tool_call["name"]
        tool_args = tool_call["args"]
        tool_call_id = tool_call["id"]

        eval_request = self._build_eval_request(tool_name, tool_args)
        start_time = time.monotonic()

        try:
            raw_response = await self._client.apost(
                self._config.authority_endpoint,
                eval_request.model_dump(),
            )
            auth_response = AuthorityEvaluationResponse(**raw_response)
        except Exception as exc:
            logger.error("Authority evaluation failed: %s", exc)
            return ToolMessage(
                content=json.dumps(
                    {
                        "status": "ERROR",
                        "tool": tool_name,
                        "message": f"Authority evaluation failed: {exc}",
                    }
                ),
                tool_call_id=tool_call_id,
            )

        if auth_response.decision == "DENY":
            logger.info("DENY: tool=%s reason=%s", tool_name, auth_response.reason_code)
            return self._make_deny_message(tool_call_id, tool_name, auth_response)

        logger.info("ALLOW: tool=%s proof_id=%s", tool_name, auth_response.proof_id)
        result = await handler(request)

        result_content = ""
        if isinstance(result, ToolMessage):
            result_content = (
                result.content if isinstance(result.content, str) else str(result.content)
            )

        total_latency = (time.monotonic() - start_time) * 1000
        proof = build_proof_artifact(
            authority_response=auth_response,
            agent_id=self._config.agent_id,
            controller_id=self._config.controller_id,
            delegation_id=self._config.delegation_id,
            tool=tool_name,
            parameters_hash=hash_parameters(tool_args),
            result_hash=hash_result(result_content),
            latency_ms=total_latency,
        )
        self._emit_proof(proof)

        return result

File: packages/js/src/authorityMiddleware.ts
import { randomUUID } from "node:crypto";
import { ToolMessage } from "@langchain/core/messages";
import { computeActionContextHashV1, hashIntentV1, hashParametersV1 } from "./actionContext.js";
import { loadPrivateKey, signAgentProofV1 } from "./agentProof.js";
import { extractOidcClientId, OidcClientCredentialsClient } from "./oidcClient.js";
import {
  buildProofArtifact,
  hashResult
} from "./proof.js";
import {
  discoverAuthority,
  ProofVerificationError,
  verifyAuthorityProof
} from "./proofVerification.js";
import type {
  ActionContext,
  ActionContextExtras,
  AuthorityEvaluationRequest,
  AuthorityEvaluationResponse,
  MiddlewareConfig,
  ProofArtifact,
  SigningKey,
  ToolArgs,
  ToolCallRequestLike
} from "./types.js";

export class NuggetsAuthorityMiddleware {
  private config: MiddlewareConfig;
  private proofList: ProofArtifact[] = [];
  private privateKeyPromise?: Promise<SigningKey>;
  private client?: OidcClientCredentialsClient;

  constructor(config: MiddlewareConfig) {
    this.config = config;
    if (config.agentPrivateKey) {
      this.privateKeyPromise = loadPrivateKey(config.agentPrivateKey);
    }
  }

  get proofs(): ProofArtifact[] {
    return [...this.proofList];
  }

  async wrapToolCall(
    request: ToolCallRequestLike,
    handler: (request: ToolCallRequestLike) => unknown | Promise<unknown>
  ): Promise<unknown> {
    const toolCall = request.tool_call;
    const toolName = toolCall.name;
    const toolArgs = toolCall.args;
    const toolCallId = toolCall.id;
    let evalRequest: AuthorityEvaluationRequest;
    try {
      evalRequest = this.buildEvalRequest(toolName, toolArgs);
    } catch (exc) {
      return this.errorMessage(toolCallId, toolName, `Action context resolution failed (${errorDetail(exc)})`);
    }
    const start = performance.now();

    let authResponse: AuthorityEvaluationResponse;
    if (this.config.testMode) {
      authResponse = this.testModeResponse();
    } else {
      try {
        const privateKey = await this.getPrivateKey();
        const [authorityIssuer] = await discoverAuthority(this.config.apiUrl);
        const actionContextHash = computeActionContextHashV1(evalRequest);
        const payload: AuthorityEvaluationRequest = {
          ...evalRequest,
          agent_proof: await signAgentProofV1(privateKey, {
            agentId: this.config.agentId,
            nonce: evalRequest.action.nonce,
            audience: authorityIssuer,
            actionContextHash
          })
        };
        const client = await this.getClient();
        const rawResponse = await client.post(
          `${this.config.apiUrl.replace(/\/+$/, "")}${this.config.authorityEndpoint}`,
          payload,
          { "Idempotency-Key": randomUUID() }
        );
        authResponse = coerceAuthorityResponse(rawResponse);
      } catch (exc) {
        return this.errorMessage(toolCallId, toolName, `Authority evaluation failed (${errorDetail(exc)})`);
      }
    }

    if (authResponse.decision === "DENY") {
      return this.denyMessage(toolCallId, toolName, authResponse);
    }

    // Verify the signed decision for BOTH ALLOW and ESCALATE before acting on it.
    const proofFailure = await this.verifyProofOrNull(
      authResponse,
      evalRequest,
      toolCallId,
      toolName
    );
    if (proofFailure) {
      return proofFailure;
    }

    if (authResponse.decision === "ESCALATE") {
      return this.escalateMessage(toolCallId, toolName, authResponse);
    }

    // ALLOW: execute + emit ProofArtifact.
    const result = await handler(request);

    // The side effect has now happened. A proof-build/emit or onProof failure
    // must NOT propagate — surfacing it as an error could cause the caller to
    // retry an already-executed, non-idempotent action.
    try {
      const resultContent = extractResultContent(result);
      const proof = buildProofArtifact({
        authorityResponse: authResponse,
        agentId: this.config.agentId,
        controllerId: this.config.controllerId,
        delegationId: this.config.delegationId,
        tool: toolName,
        parametersHash: evalRequest.action.parameters_hash,
        resultHash: hashResult(resultContent),
        latencyMs: performance.now() - start,
        intentHash: evalRequest.action.intent_hash ?? null,
        testMode: this.config.testMode
      });
      await this.emitProof(proof);
    } catch {
      // Executed; proof persistence/callback failed. Swallow so the completed
      // execution is not reported as a failure (no logger in this package).
    }
    return result;
  }

  async awrapToolCall(
    request: ToolCallRequestLike,
    handler: (request: ToolCallRequestLike) => unknown | Promise<unknown>
  ): Promise<unknown> {
    return this.wrapToolCall(request, handler);
  }

  buildEvalRequest(toolName: string, toolArgs: ToolArgs): AuthorityEvaluationRequest {
    const intent = this.config.intentResolver?.(toolName, toolArgs) ?? null;
    const extras = validateActionContextExtras(this.config.actionContextResolver?.(toolName, toolArgs));
    const defaultTarget = typeof toolArgs.target === "undefined" ? toolName : String(toolArgs.target);
    const parametersHash = hashParametersV1(toolArgs);
    const timestamp = new Date().toISOString();
    const intentHash = intent === null ? null : hashIntentV1(intent);
    const action: ActionContext = {
      tool: toolName,
      target: extras?.target ?? defaultTarget,
      parameters_hash: parametersHash,
      intent,
      intent_hash: intentHash,
      timestamp,
      nonce: randomUUID()
    };
    if (typeof extras?.amount_minor === "number") action.amount_minor = extras.amount_minor;
    if (typeof extras?.currency === "string") action.currency = extras.currency;
    return {
      agent_id: this.config.agentId,
      controller_id: this.config.controllerId,
      delegation_id: this.config.delegationId,
      action
    };
  }

  private async getPrivateKey(): Promise<SigningKey> {
    if (!this.privateKeyPromise) {
      throw new Error("agentPrivateKey required when testMode is false");
    }
    return this.privateKeyPromise;
  }

  private async getClient(): Promise<OidcClientCredentialsClient> {
    if (this.client) {
      return this.client;
    }
    if (!this.config.oidcIssuerUrl) {
      throw new Error("oidcIssuerUrl required when testMode is false");
    }
    const privateKey = await this.getPrivateKey();
    this.client = new OidcClientCredentialsClient({
      issuerUrl: this.config.oidcIssuerUrl,
      clientId: extractOidcClientId(this.config.agentId),
      privateKey,
      scope: this.config.authorityScope,
      resource: this.config.resolvedAuthorityAudience()
    });
    return this.client;
  }

  private async verifyProofOrNull(
    authResponse: AuthorityEvaluationResponse,
    evalRequest: AuthorityEvaluationRequest,
    toolCallId: string,
    toolName: string
  ): Promise<ToolMessage | null> {
    if (this.config.testMode || !this.config.verifyProofs) {
      return null;
    }
    try {
      const [issuer, jwksUri] = await discoverAuthority(this.config.apiUrl);
      await verifyAuthorityProof({
        signature: authResponse.signature,
        expected: {
          decision: authResponse.decision,
          proof_id: authResponse.proof_id,
          agent_id: this.config.agentId,
          controller_id: this.config.controllerId,
          aud: this.config.agentId,
          action_context_version: 1,
          action_context_hash: computeActionContextHashV1(evalRequest),
          constraints_evaluated: authResponse.constraints_evaluated ?? []
        },
        issuer,
        jwksUri
      });
      return null;
    } catch (exc) {
      if (exc instanceof ProofVerificationError || exc instanceof Error) {
        return this.proofFailureMessage(exc, toolCallId, toolName, authResponse.proof_id);
      }
      return this.proofFailureMessage(new Error(String(exc)), toolCallId, toolName, authResponse.proof_id);
    }
  }

  private testModeResponse(): AuthorityEvaluationResponse {
    return {
      decision: "ALLOW",
      proof_id: `test-${randomUUID()}`,
      signature: "test-mode-unverifiable",
      reason_code: null,
      constraints_evaluated: ["test_mode"]
    };
  }

  private denyMessage(toolCallId: string, toolName: string, response: AuthorityEvaluationResponse): ToolMessage {
    return new ToolMessage({
      content: JSON.stringify({
        status: "DENIED",
        tool: toolName,
        reason_code: response.reason_code ?? null,
        proof_id: response.proof_id,
        message: `Authority check denied execution of '${toolName}'${
          response.reason_code ? `: ${response.reason_code}` : ""
        }`
      }),
      tool_call_id: toolCallId
    });
  }

  private escalateMessage(toolCallId: string, toolName: string, response: AuthorityEvaluationResponse): ToolMessage {
    return new ToolMessage({
      content: JSON.stringify({
        status: "PENDING_APPROVAL",
        tool: toolName,
        approval_id: response.approval_id ?? null,
        reason_code: response.reason_code ?? null,
        proof_id: response.proof_id,
        signature: response.signature,
        constraints_evaluated: response.constraints_evaluated ?? [],
        message: `Execution of '${toolName}' is pending human approval${
          response.approval_id ? ` (approval ${response.approval_id})` : ""
        }. This is not an error. The ESCALATE decision's signature is verified; 'approval_id' is a server-issued handle, not part of the signed receipt. Present or poll the approval to continue.`
      }),
      tool_call_id: toolCallId
    });
  }

  private errorMessage(toolCallId: string, toolName: string, message: string): ToolMessage {
    return new ToolMessage({
      content: JSON.stringify({
        status: "ERROR",
        tool: toolName,
        message
      }),
      tool_call_id: toolCallId
    });
  }

  private proofFailureMessage(
    exc: Error,
    toolCallId: string,
    toolName: string,
    proofId: string
  ): ToolMessage {
    return new ToolMessage({
      content: JSON.stringify({
        status: "DENIED",
        tool: toolName,
        reason_code: "PROOF_VERIFICATION_FAILED",
        proof_id: proofId,
        message: `Authority proof for '${toolName}' failed verification: ${exc.message}`
      }),
      tool_call_id: toolCallId
    });
  }

  private async emitProof(proof: ProofArtifact): Promise<void> {
    this.proofList.push(proof);
    await this.config.onProof?.(proof);
  }
}

/**
 * A non-sensitive descriptor of a thrown value for user-facing ERROR messages.
 * Returns the error's *constructor* name (or the primitive type) — never the
 * message or the instance `.name`, both of which are writable and can carry
 * resolver/tool-arg data that would otherwise reach the LLM/user. The
 * constructor name reflects the (statically defined) class, so it is safe.
 */
function errorDetail(exc: unknown): string {
  if (exc instanceof Error) return exc.constructor?.name || "Error";
  return typeof exc;
}

function validateActionContextExtras(
  extras: ActionContextExtras | undefined
): ActionContextExtras | undefined {
  if (extras === undefined) return undefined;
  if (typeof extras !== "object" || extras === null) {
    throw new Error("actionContextResolver must return an object or undefined");
  }
  if (extras.target !== undefined && (typeof extras.target !== "string" || extras.target.trim().length === 0)) {
    throw new Error("actionContextResolver target must be a non-empty string");
  }
  const hasAmount = extras.amount_minor !== undefined;
  const hasCurrency = extras.currency !== undefined;
  if (hasAmount !== hasCurrency) {
    throw new Error("actionContextResolver must supply amount_minor and currency together, or neither");
  }
  if (hasAmount && (!Number.isSafeInteger(extras.amount_minor) || (extras.amount_minor as number) < 0)) {
    throw new Error("actionContextResolver amount_minor must be a non-negative safe integer");
  }
  if (hasCurrency && !/^[A-Z]{3}$/.test(extras.currency as string)) {
    throw new Error("actionContextResolver currency must be an uppercase ISO-4217 code");
  }
  return extras;
}

function coerceAuthorityResponse(value: unknown): AuthorityEvaluationResponse {
  if (!value || typeof value !== "object") {
    throw new Error("authority response is not an object");
  }
  const response = value as Record<string, unknown>;
  if (response.decision !== "ALLOW" && response.decision !== "DENY" && response.decision !== "ESCALATE") {
    throw new Error("authority response missing decision");
  }
  if (typeof response.proof_id !== "string") {
    throw new Error("authority response missing proof_id");
  }
  if (typeof response.signature !== "string") {
    throw new Error("authority response missing signature");
  }
  return {
    decision: response.decision,
    proof_id: response.proof_id,
    signature: response.signature,
    reason_code: typeof response.reason_code === "string" ? response.reason_code : null,
    approval_id:
      typeof response.approval_id === "string" || typeof response.approval_id === "number"
        ? response.approval_id
        : null,
    constraints_evaluated: Array.isArray(response.constraints_evaluated)
      ? response.constraints_evaluated.filter((item): item is string => typeof item === "string")
      : []
  };
}

function extractResultContent(result: unknown): string {
  if (result && typeof result === "object" && "content" in result) {
    const content = (result as { content: unknown }).content;
    return typeof content === "string" ? content : JSON.stringify(content);
  }
  return "";
}
